### PR TITLE
feat: popover portals

### LINKED ISSUE #225: Use portals to position popovers above the rest of the DOM
Current behavior works great 99% of the time, but using portals wouldn't be a huge change and would ensure no overlap with other components

### PR DESCRIPTION
**Pull request by alexpaxton**
Closes #225

tldr; rebuild `Popover` from the ground up

### Why

- When using the current `Popover` within a stacking context the positioning gets really wonky and frustrating
- Wrapping the trigger element in a `<div />` essentially negates the layout styles the trigger element had, which requires additional styles to fix
- `Popover` is not ensured to be above all other UI elements
- `Popover` tends to re-render excessively

### Changes

- Instead of wrapping a child now you pass in a `ref` to the triggering element
  - This prevents layout disruption
- `Popover` appears as a child of `<body />` via a react portal
- Add `enableDefaultStyles` prop to `Popover`
  - This is true by default and saves you the trouble of having to style all your `Popovers`
  - When turned off the popover dialog is free of styles so you can get creative
- Introduce `ReflessPopover` as a composed version of `Popover`
  - The main difference is (you guessed it) that it does not require a `ref` to work
  - Mimics the behavior of the existing `Popover` component (pass in trigger element as a child and the child is wrapped)

### Screenshots

Screen Shot 2019-09-11 at 2 02 50 PM

### Checklist
Check all that apply

- [x] Updated documentation to reflect changes
- [x] Added entry to top of Changelog with link to PR (not issue)
- [x] Tests pass
- [x] Peer reviewed and approved
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: src/Components/Popover/PopoverDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+
+  private get contentsClassName(): string {
+    const {enableDefaultStyles} = this.props
+
+    return classnames('cf-popover--contents', {
+      'cf-popover--contents__default-styles': enableDefaultStyles,
+    })
+  }
+
+  private calculateDialogPosition = (): PopoverPosition => {
```
Comment (chnn) [inline src/Components/Popover/PopoverDialog.tsx]: I think that this method (`calculateDialogPosition`) along with `isDialogFlush` and `popoverStyles` should be utility functions rather than methods on this class.

As functions their dependencies are explicit (arguments), but as methods they have access to everything on `this`. I think when dependencies are made explicit, it opens up opportunities to refactor functions. And these are really gnarly, so any refactor here would be a win imo.

File: src/Components/Popover/Popover.tsx [hunk 2]
```diff
@@ -465,45 +195,126 @@ export class Popover extends Component<PopoverProps, State> {
...
-
     this.setState({expanded: true})
   }
 
   private handleHideDialog = (): void => {
     this.setState({expanded: false})
+  }
+
+  private handleCreatePortalElement = (): void => {
+    const portalExists = document.getElementById('cf-popover-portal')
```
Comment (mavarius) [inline src/Components/Popover/Popover.tsx]: The ref could also work for this

File: src/Components/Popover/Popover.tsx [hunk 3]
```diff
@@ -89,6 +99,23 @@ export class Popover extends Component<PopoverProps, State> {
   }
 
   public componentDidUpdate(prevProps: PopoverProps) {
+    if (!this.triggerReceivedHandlers) {
+      this.triggerReceivedHandlers = false
+
+      this.props.triggerRef.current.addEventListener(
```
[the remark this answers is not in the document]
Comment (alexpaxton) [reply src/Components/Popover/Popover.tsx]: I think it could be fine-tuned a bit. I have this extra step in `componentDidUpdate` because when the component mounts the trigger ref doesn't exist yet and the handlers don't get added. Might make sense to just add the handlers on the first `componentDidUpdate` rather than `componentDidMount`

File: src/Components/Popover/Popover.tsx [hunk 4]
```diff
@@ -465,45 +195,126 @@ export class Popover extends Component<PopoverProps, State> {
...
+    portalElement.setAttribute('class', 'cf-popover-portal')
+    portalElement.setAttribute('id', 'cf-popover-portal')
+
+    document.body.appendChild(portalElement)
+  }
+
+  private handleDestroyPortalElement = (): void => {
+    const portalElement = document.getElementById('cf-popover-portal')
+
+    if (portalElement) {
```
[the remark this answers is not in the document]
Comment (alexpaxton) [reply src/Components/Popover/Popover.tsx]: that's the thinking I need in this world

**Comment by alexpaxton** [comment]
@chnn I made some changes based on your feedback:

- Passed the `triggerRef` into the dialog component instead of holding the rect in state
- Moved the positioning logic out of `PopoverDialog` and into a utils file
- Moved the window event listeners into `PopoverDialog`
- There is now a unique portal element for each popover
- Forced a single re-render right after `Popover` mounts so that `triggerRef.current` can be rendered in the parent and not get passed into `Popover` as `null`

The whole thing feels way smoother and snappier now, which tells me there is way less re-rendering going on.
